Date enhancements and support for Unix Timestamps

galenandrew opened the pull request:
_Revised_
--

This PR includes the following changes:
- Support for decimals in `date()` validation
- New constraint for `date.timestamp()` that supports both _javascript_ and _unix_ timestamps
- Updated Tests, API documentation, and added date example


_Note: Because javascript timestamps are millisecond-based and unix timestamps are second-based, the `timestamp()` constraint accepts an optional `type` parameter to ensure proper date conversion for the given input value._

closes #789

Marsup commented on lib/date.js:
```diff
@@ -36,34 +45,56 @@ Hoek.inherits(internals.Date, Any);
+            /^[+-]?\d+(\.\d+)?$/.test(value)) {
 
-            value = parseInt(value, 10);
+            value = Number(value);
```
Why ?

galenandrew replied:
I changed this to the Number constructor because `parseInt` drops the decimals. After [reviewing performance](https://jsperf.com/number-vs-parseint-vs-plus) I've updated it to use `parseFloat`.

Marsup commented on lib/date.js:
```diff
@@ -36,34 +45,56 @@ Hoek.inherits(internals.Date, Any);
 internals.Date.prototype._base = function (value, state, options) {
 
+    if (this._unit) {
+        internals.isValidUnit(this._unit);
```
Can't crash _base like that. I'd prefer you override .unit() and call any's implementation rather than that late check.

Marsup commented on lib/date.js:
```diff
@@ -151,6 +182,32 @@ internals.Date.prototype.iso = function () {
+        internals.isValidUnit(obj._unit);
+    }
+
+    obj._flags.timestamp = true;
```
Stick to the format field, I don't like the fact that flag can conflict each other and introduce priority.

Marsup commented on lib/date.js:
```diff
@@ -151,6 +182,32 @@ internals.Date.prototype.iso = function () {
     return obj;
 };
 
+internals.Date.prototype.timestamp = function (unit) {
```
There's already .unit(), I'm not sure that parameter adds much.

Marsup commented:
Thinking about it, can't we just use the existing unit without adding any new api ?

galenandrew commented:
@Marsup PR has been updated and downstream merged with your recent commits to `master`.

I've removed the approach that uses `unit` and simplified the new `timestamp` constraint implementation.

Marsup commented:
So what happened to my last comment ?

galenandrew commented:
I'm not sure I understand what you mean? Please explain.

Marsup commented:
I don't think a new API is needed, just using the existing unit.

galenandrew commented:
I agree that replicating the `unit` API within the date type is not needed, however I respectfully disagree about the need for a new API for timestamps. The date type has an explicit `iso` constraint and should have an explicit `timestamp` constraint as well.

This PR is clean, functional and has 100% test coverage.

galenandrew commented:
If you have objective reasoning for a different approach on handling timestamps let's discuss it. For instance, getting rid of the `type` parameter in the `timestamp` API and using the built in `unit` to handle the multiplier.

Marsup commented:
`Joi.date().timestamp('unix')` is essentially the same as `Joi.date().unit('s')` except you're adding a new API and a new concept of "unix timestamp" that not everyone might be familiar with. `Joi.date()` just defaults to `unit('ms')` when it's a number.

ISO is a strict format applied on a string hence its special treatment, a number is not timestamp by itself, any number is a timestamp. Actually, if you want a strict validation on that, I'm just now seeing that this could already be achieved by a `Joi.date().format('X')` or `Joi.date().format('x')`.

> using the built in unit to handle the multiplier

I do mean that the unit should translate to a multiplier.

FWIW I'm not questioning the quality of the PR here, just the content.

kanongil commented:
@Marsup I am pretty sure it is a bad idea to change the parsing of a date depending on the `.unit()` value, given that it is only for annotation purposes.

Also, `Joi.date().format('X')` won't work for fractional timestamps.

Marsup commented:
It's been for annotation *so far*, is there any valid reason not to use it as it already exists ?
Fractional as in `x/y` ? Moment seems to support floats according to their docs, I didn't test.

kanongil commented:
Fractional, as in 123.45, which Joi never passes on to moment.

Marsup commented:
Oh yeah, but with this patch it would.

galenandrew commented:
While using unit to set the timestamp multiple is a valid approach, what are the reasons we should _not_ add an explicit `timestamp` constraint?

To me, `Joi.date().timestamp()` is more descriptive about the function of this constraint vs `Joi.date().unit('s')`.

Marsup commented:
I don't mind the timestamp per se, the flag behind it seems like a useless duplication of unit though. As common ground I'd propose timestamp takes whatever it can and translate it to unit, OK for everyone ?

galenandrew commented:
That can work. So to confirm, is the following acceptable? If so, I'll update the PR accordingly.

> The timestamp constraint will apply a numeric regex (`/^[+-]?\d+(\.\d+)?$/`) and any argument passed will set the existing unit flag (which will be used to control the multiplier).

So for instance…
`Joi.date().timestamp('seconds')` or `Joi.date().timestamp().unit('seconds')` will explicitly test for numeric values AND set the existing unit to seconds which will as a result trigger the multiplier. In other words, a _stricter_ validation (eg non-timestamp values will fail).

`Joi.date().unit('seconds')` will trigger the multiplier, but not force the numeric values regex. In other words, a _less strict_ validation (eg non-timestamp values will _not_ fail but the multiplier may not be used based on the value format).

kanongil commented:
I'd still argue against changing the meaning of `unit(…)`. If you decide to go ahead, remember to flag this as a breaking change.

galenandrew commented:
Honestly, I'm leaning that way as well. Avoiding unit feels cleaner and follows a better separation of concerns. "Piggy-backing" on unit introduces cross-over between the APIs and potential complications down the road if the functionality/purpose of one changes.

To me, the question should come down to the `timestamp(...)` parameter using a concept of **type** vs **time segment** – unix/javascript vs seconds/milliseconds. Type feels more descriptive, but time segment seems like it would allow for greater flexibility with the multiplier and value conversion.

Marsup commented:
Back-channeled a bit with @kanongil, let's stick to the current form, gonna review now.

Marsup commented on lib/date.js:
```diff
@@ -23,6 +23,12 @@ internals.isIsoDate = (() => {
         return date && (date.toString() === isoString);
     };
 })();
+internals.isTimestampAllowed = function (unit) {
```
Inline that, it's only used once.

Marsup commented on lib/date.js:
```diff
@@ -32,37 +38,50 @@ internals.Date = function () {
     return result;
 };
 
+internals.toDate = function (value, obj) {
```
Provide the flags directly, makes for a much cleaner code afterwards.

Marsup commented on lib/date.js:
```diff
@@ -32,37 +38,50 @@ internals.Date = function () {
-internals.toDate = function (value, format) {
+    obj = obj || { _flags: {} };
+    const format = obj._flags.format;
+    const multiplier = obj._flags.timestamp === 'unix' ? 1000 : null;
```
Could also default to 1 so that we don't need another ternary.

Marsup commented on lib/date.js:
```diff
@@ -150,6 +166,14 @@ internals.Date.prototype.iso = function () {
+
+    const obj = this.clone();
+    obj._flags.timestamp = type ? type : 'javascript';
+    internals.isTimestampAllowed(obj._flags.timestamp);
```
Make the check before cloning.

Marsup commented on test/date.js:
```diff
@@ -32,6 +32,16 @@ describe('date', () => {
+
+        const schema = Joi.date();
+        Helper.validate(schema, [
+            [Infinity, false],
```
From now on I'm trying to always test error messages as well (https://github.com/hapijs/joi/blob/master/test/any.js#L1463).

Marsup commented on test/date.js:
```diff
@@ -304,6 +316,83 @@ describe('date', () => {
+                expect(() => {
+
+                    Joi.date().timestamp('not allowed');
+                }).to.throw(Error, new RegExp('"type" must be one of'));
```
Prefer the // form.

Marsup commented on test/date.js:
```diff
@@ -304,6 +316,83 @@ describe('date', () => {
+                done();
+            });
+
+            it('validates timestamps with a friendly error message', (done) => {
```
Not sure this test is needed once previous tests also validate the error message.

Marsup commented on test/date.js:
```diff
@@ -304,6 +316,83 @@ describe('date', () => {
+                done();
+            });
+        });
+
```
Missing a test asserting correct toggling of the flag (`Joi.date().timestamp().timestamp('unix')`)

Marsup replied:
It would be nice to check for more ambiguous errors such as number-like (`1E3`, '123A`, `0xA`, ...)

galenandrew commented:
Thanks for the review and line notes @Marsup!! I'll work on these changes and let you know when it's updated.

galenandrew commented:
@Marsup, PR is updated. Good call on adding some additional checks for ambiguous values. I've also updated the `date.base` language to be timestamp agnostic.

When all looks good let me know and I'll rebase down to a single commit so it keeps the history short.

Marsup commented on lib/date.js:
```diff
@@ -150,6 +159,19 @@ internals.Date.prototype.iso = function () {
 
+internals.Date.prototype.timestamp = function (type) {
+
+    type = type ? type : 'javascript';
```
type = type || 'javascript'

galenandrew replied:
Good call…will update.

Marsup commented on lib/language.js:
```diff
@@ -54,10 +54,14 @@ exports.errors = {
     },
     date: {
-        base: 'must be a number of milliseconds or valid date string',
+        base: 'must be a valid date string or timestamp',
```
What was wrong with that ? It's still valid even with that change.

galenandrew replied:
The updated message is more accurate and explicit.

Marsup replied:
Seems more explicit for someone unaware of what a timestamp is no ?

galenandrew replied:
Maybe explicit is not the right term. _"number of milliseconds"_ does not apply to all timestamp types which is why I updated it to read _"timestamp"_ instead.

I would argue that developers who are using date validation are going to generally know what a timestamp is. If not, it is their responsibility to learn what a timestamp is…not the error messages.

galenandrew replied:
If they use the explicit `timestamp(...)` validator, the error message returned is explicit to that type stating _"number of milliseconds"_ or _"number of seconds"_.

Marsup replied:
A timestamp might not always be in milliseconds but javascript expects that and so does a base joi date, otherwise it'd be a `timestamp.unix` error.

Marsup replied:
Maybe `timestamp.javascript` should not even exist since I think it should be the same error message.

galenandrew replied:
I'll agree to disagree here. Since you're the maintainer, ultimately it's your decision. If this is that important to you I'll update it and rebase so you'll have a clean merge. What do you want to do?

Marsup replied:
Since the javascript mode is essentially keeping the behavior we already had, let's keep the errors identical.

galenandrew replied:
done

Marsup commented on test/date.js:
```diff
@@ -304,6 +316,116 @@ describe('date', () => {
+                    [1.452126061677e+12, true],
+                    [1E3, true],
+                    ['1E3', true],
+                    [',', false],
```
Can you also check error messages here please ?

Marsup commented on test/date.js:
```diff
@@ -32,6 +32,16 @@ describe('date', () => {
+
+        const schema = Joi.date();
+        Helper.validate(schema, [
+            [Infinity, false, null, /.* must be a valid date string/],
```
Those .* are not necessary.

Marsup commented on test/date.js:
```diff
@@ -304,6 +316,116 @@ describe('date', () => {
+                done();
+            });
+
+            it('validates timestamps with a friendly error message', (done) => {
```
I really feel like this test should be merged with the 2 first functions of timestamp tests.

galenandrew replied:
Originally I was simply following the structure of the other tests in this file. On this last update, removing this test block did not pass code coverage. I'll see what I can do here, but tests are tests.

galenandrew commented:
PR has been updated

galenandrew commented:
PR has been updated and rebased

Marsup commented:
Thanks for the hard work.

galenandrew commented:
You're welcome!! Thanks for the feedback and reviewing my pull request!

AdrieanKhisbe commented:
:+1:
